File: torch/basic.py
import torch

from utils import GPU, simple_timer

N = 64

def create_zero_gpu():
    dev = GPU
    
    for _ in range(0, N):
        t = torch.zeros(1024, 8192, dtype=torch.float32, device=dev)

def create_ones_gpu():
    dev = GPU
    
    for _ in range(0, N):
        t = torch.ones(1024, 8192, dtype=torch.float32, device=dev)

def create_full_gpu():
    dev = GPU
    
    for i in range(0, N):
        t = torch.full((1024, 8192), i, dtype=torch.float32, device=dev)


ctg = simple_timer(create_zero_gpu, GPU)
print(f"Zeroes: {ctg}s {ctg / 10000}s/iter")

ctg = simple_timer(create_ones_gpu, GPU)
print(f"Ones: {ctg}s {ctg / 10000}s/iter")

ctg = simple_timer(create_full_gpu, GPU)
print(f"Full: {ctg}s {ctg / 10000}s/iter")

File: src/basic.rs
use candle_core::{DType, Device, Tensor};

const N: usize = 64;

pub fn create_zero_gpu(dev: &Device) {
    for _ in 0 .. N {
        let _ = Tensor::zeros((1024, 8192), DType::F32, dev);
    }
}

pub fn create_ones_gpu(dev: &Device) {
    for _ in 0 .. N {
        let _ = Tensor::ones((1024, 8192), DType::F32, dev);
    }
}

pub fn create_full_gpu(dev: &Device) {
    for i in 0 .. N {
        let _ = Tensor::full( i as f32, (1024, 8192), dev);
    }
}

#[cfg(test)]
mod tests {
    use crate::{select_device, simple_timer};

    use super::*;

    #[test]
    fn test_create_tensor() {
        let dev = select_device().unwrap();
        let t = simple_timer(&dev, create_zero_gpu);
        println!("Zeroes: {t}s {}s/iter", t/ N as f32);

        let t = simple_timer(&dev, create_ones_gpu);
        println!("Ones: {t}s {}s/iter", t/ N as f32);

        let t = simple_timer(&dev, create_full_gpu);
        println!("Full: {t}s {}s/iter", t/ N as f32);
    }
}